### PR TITLE
Add snowflake integration

### PR DESCRIPTION
**Pull request by BinaryFissionGames**
#### PR Description
* Adds the [snowflake exporter](https://github.com/grafana/snowflake-prometheus-exporter) as an integration.

#### PR Checklist

- [x] CHANGELOG updated
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: docs/sources/configuration/integrations/integrations-next/snowflake-config.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    # Autoscrape interval and timeout. Defaults are inherited from the global
+    # section of the top-level metrics config.
+    [scrape_interval: <duration> | default = <metrics.global.scrape_interval>]
+    [scrape_timeout: <duration> | default = <metrics.global.scrape_timeout>]
+
+  # Integration instance name. 
+  # The default value for this integration is the configured account_name.
+  [instance: <string> | default = <account_name>]
+
+  # The account name of the snowflake account to monitor
```
Comment (karengermond) [inline docs/sources/configuration/integrations/integrations-next/snowflake-config.md:37]: Should be consistent with periods.

File: docs/sources/configuration/integrations/integrations-next/snowflake-config.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  account_name: <string>
+
+  # Username for the database user used to scrape metrics.
+  username: <string>
+
+  # Password for the database user used to scrape metrics
```
Comment (karengermond) [inline docs/sources/configuration/integrations/integrations-next/snowflake-config.md:43]:
```suggestion
  # Password for the database user used to scrape metrics.
```

File: docs/sources/configuration/integrations/integrations-next/snowflake-config.md (hunk 1 above)
Comment (karengermond) [inline docs/sources/configuration/integrations/integrations-next/snowflake-config.md:37]:
```suggestion
  # The account name of the snowflake account to monitor.
```

File: docs/sources/configuration/integrations/snowflake-config.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+The `snowflake_config` block configures the `snowflake` integration,
+which is an embedded version of
+[`snowflake-prometheus-exporter`](https://github.com/grafana/snowflake-prometheus-exporter). This allows the collection of [Snowflake](https://www.snowflake.com/) metrics.
+
+Full reference of options:
+
+```yaml
+  # Enables the apache_http integration, allowing the Agent to automatically
+  # collect metrics for the specified snowflake account.
```
Comment (karengermond) [inline docs/sources/configuration/integrations/snowflake-config.md:17]: Is it enabling the apache_http integration?  I thought this was just the snowflake integration but maybe it needs that integration, too?

File: docs/sources/configuration/integrations/snowflake-config.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  # self-scraped. Overrides inferred values.
+  #
+  # The default value for this integration is the configured account_name.
+  [instance: <string>]
+
+  # Automatically collect metrics from this integration. If disabled,
+  # the snowflake integration will be run but not scraped and thus not
+  # remote-written. Metrics for the integration will be exposed at
+  # /integrations/snowflake/metrics and can be scraped by an external
+  # process.
```
Comment (karengermond) [inline docs/sources/configuration/integrations/snowflake-config.md:30]:
```suggestion
  # Automatically collect metrics from this integration. If disabled,
  # the snowflake integration is run but not scraped and thus not
  # remote-written. Metrics for the integration are exposed at
  # /integrations/snowflake/metrics and can be scraped by an external
  # process.
```

File: docs/sources/configuration/integrations/snowflake-config.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  # The timeout before considering the scrape a failure. Defaults to
+  # prometheus.global.scrape_timeout.
+  [scrape_timeout: <duration> | default = <global_config.scrape_timeout>]
+
+  # Allows for relabeling labels on the target.
+  relabel_configs:
+    [- <relabel_config> ... ]
+
+  # Relabel metrics coming from the integration, allowing to drop series
+  # from the integration that you don't care about.
```
Comment (karengermond) [inline docs/sources/configuration/integrations/snowflake-config.md:46]:
```suggestion
  # Relabel metrics coming from the integration, lets you drop series
  # that you don't care about from the integration.
```

File: docs/sources/configuration/integrations/snowflake-config.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  metric_relabel_configs:
+    [ - <relabel_config> ... ]
+
+  # How frequent to truncate the WAL for this integration.
```
Comment (karengermond) [inline docs/sources/configuration/integrations/snowflake-config.md:50]:
```suggestion
  # How frequently the WAL is truncated for this integration.
```

File: docs/sources/configuration/integrations/snowflake-config.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  [wal_truncate_frequency: <duration> | default = "60m"]
+
+  #
+  # Exporter-specific configuration options
+  #
+     
+  # The account name of the snowflake account to monitor
```
Comment (karengermond) [inline docs/sources/configuration/integrations/snowflake-config.md:57]:
```suggestion
  # The account name of the snowflake account to monitor.
```

File: docs/sources/configuration/integrations/snowflake-config.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  account_name: <string>
+
+  # Username for the database user used to scrape metrics.
+  username: <string>
+
+  # Password for the database user used to scrape metrics
```
Comment (karengermond) [inline docs/sources/configuration/integrations/snowflake-config.md:63]:
```suggestion
  # Password for the database user used to scrape metrics.
```

**Review by karengermond — COMMENTED** [review]
LGTM -- a few comments.

File: docs/sources/configuration/integrations/snowflake-config.md (hunk 3 above)
Comment (BinaryFissionGames) [reply docs/sources/configuration/integrations/snowflake-config.md]: Good catch; I had copied to common settings from the apache_http doc and missed this!

**Comment by BinaryFissionGames** [comment]
Thanks for the review @karengermond! I've made the suggested changes.

File: CHANGELOG.md [hunk 9]
```diff
@@ -56,6 +56,7 @@ Main (unreleased)
   - `loki.process` receives log entries from other `loki` components and runs
     one or more processing stages. (@tpaschalis)
 
+- Integrations: Introduce the `snowflake` integration. (@binaryfissiongames)
```
Comment (marctc) [inline CHANGELOG.md:59]: nit: new line in between

**Review by marctc — APPROVED** [review]
LGTM

**Comment by mattdurham** [comment]
Looks like updating the S3 lib broke a test, and verified that main works. Will try to recreate on your branch

**Review by mattdurham — CHANGES_REQUESTED** [review]
I think the s3 test is flaky due to timing. Can we increase the timeout in `s3_test.go` to 1 second?  `time.Sleep(1000 * time.Second)`

**Comment by mattdurham** [comment]
Oh no, glad you went with what I asked as opposed to my code snippet I posted, 1,000 seconds would have been a long time!

**Comment by BinaryFissionGames** [comment]
Hm, it didn't seem to work, unfortunately. 

The test does seem to pass locally on my MacBook (when I remove the build constraing), so it's interesting that it doesn't work in CI.

**Comment by mattdurham** [comment]
Agreed, I cant reproduce the error. These type of tests sometimes fails on the drone runners since those are constrained environments but this is consistent.

**Comment by mattdurham** [comment]
Looking into the drone build debug logs to see if anything stands out. If it was breaking on `main` then I wouldn't be concerned or at least could blame drone. But in this case main tests pass for other PRs and this did update s3 with a patch version.

**Comment by mattdurham** [comment]
This test has been flaky in the past, one of the reasons I only had it run on linux. Go ahead and remove it and I will approve the PR.

**Comment by BinaryFissionGames** [comment]
@mattdurham I removed the offending test;
For future reference, it looks like it was trying to reach out to EC2 (I know very little about AWS, so not sure what to make of it).
`operation error S3: GetObject, failed to sign request: failed to retrieve credentials: failed to refresh cached credentials, no EC2 IMDS role found, operation error ec2imds: GetMetadata, http response error StatusCode: 404, request to EC2 IMDS failed`

**Comment by mattdurham** [comment]
My thought is updating the library changed something subtly and the Grafana drone containers may have some credentials that are getting confused between the fake s3 and actual object storage (GCP , AWS).

**Review by mattdurham — APPROVED** [review]

**Comment by mattdurham** [comment]
Thank you so much for your submission and quick turn around on requests.

**Comment by BinaryFissionGames** [comment]
Same to you, thanks for the quick turnaround on responses and helping debug that issue!

**Pull request merged**